jupyter_client/tests/utils.py

```diff
@@ -181,3 +186,101 @@ class AsyncKernelManagerWithCleanup(AsyncKernelManager):
     def cleanup(self, connection_file=True):
         super().cleanup(connection_file=connection_file)
         self.which_cleanup = 'cleanup'
+
+
+class SyncMKMSubclass(RecordCallMixin, MultiKernelManager):
+
+    def _kernel_manager_class_default(self):
+        return 'jupyter_client.tests.utils.SyncKernelManagerSubclass'
+
+    def get_kernel(self, kernel_id):
+        self.record('get_kernel')
+        return super().get_kernel(kernel_id)
+
+    def remove_kernel(self, kernel_id):
+        self.record('remove_kernel')
+        return super().remove_kernel(kernel_id)
+
+    def start_kernel(self, kernel_name=None, **kwargs):
+        self.record('start_kernel')
+        return super().start_kernel(kernel_name=kernel_name, **kwargs)
+
+    def shutdown_kernel(self, kernel_id, now=False, restart=False):
+        self.record('shutdown_kernel')
+        return super().shutdown_kernel(kernel_id, now=now, restart=restart)
+
+    def restart_kernel(self, kernel_id, now=False):
+        self.record('restart_kernel')
+        return super().restart_kernel(kernel_id, now=now)
+
+    def interrupt_kernel(self, kernel_id):
+        self.record('interrupt_kernel')
+        return super().interrupt_kernel(kernel_id)
+
+    def request_shutdown(self, kernel_id, restart=False):
+        self.record('request_shutdown')
+        return super().request_shutdown(kernel_id, restart=restart)
+
+    def finish_shutdown(self, kernel_id, waittime=None, pollinterval=0.1):
+        self.record('finish_shutdown')
+        return super().finish_shutdown(kernel_id, waittime=waittime, pollinterval=pollinterval)
+
+    def cleanup_resources(self, kernel_id, restart=False):
+        self.record('cleanup_resources')
+        super().cleanup_resources(kernel_id, restart=restart)
+
+    def shutdown_all(self, now=False):
+        self.record('shutdown_all')
+        return super().shutdown_all(now=now)
+
+
+class AsyncMKMSubclass(RecordCallMixin, AsyncMultiKernelManager):
+    """Used to test subclass hierarchies to ensure methods are called when expected.
+
+       This class is also used to test deprecation "routes" that are determined by superclass'
+       detection of methods.
+
+       This class represents a current subclass that overrides "interesting" methods of AsyncKernelManager.
+    """
+    def _kernel_manager_class_default(self):
+        return 'jupyter_client.tests.utils.AsyncKernelManagerSubclass'
+
+    def get_kernel(self, kernel_id):
+        self.record('get_kernel')
+        return super().get_kernel(kernel_id)
+
+    def remove_kernel(self, kernel_id):
+        self.record('remove_kernel')
+        return super().remove_kernel(kernel_id)
+
+    async def start_kernel(self, kernel_name=None, **kwargs):
+        self.record('start_kernel')
+        return await super().start_kernel(kernel_name=kernel_name, **kwargs)
```

Comment:
It looks like we could have a common base class for `SyncMKMSubclass` and `AsyncMKMSubclass`, rather than duplicating all the methods. All the async methods can be regular functions, they will just return a result (sync) or an awaitable (async).
You could also use a `mkm_method` decorator as in:
https://github.com/jupyter/jupyter_client/blob/26a16c0c91e245f7403aa27a812fee5e905d2964/jupyter_client/multikernelmanager.py#L27-L40

Reply:
Thanks @davidbrochart.  I'm having trouble visualizing the common base class approach that applies to both sync/async subclasses, could you please give an example?

Regarding the decorator, I'm not very familiar with this kind of thing, would it resemble something like this:
```python

def mkm_recorder(f):
    def wrapped(self, *args, **kwargs): 
         # record this call
         self.record(f.__name__) 
         method = getattr(super(), f.__name__) 
         # call the superclass method 
         r = method(*args, **kwargs) 
         return r 
     return wrapped 

class AsyncMKMSubclass(RecordCallMixin, AsyncMultiKernelManager):
    @mkm_recorder
    async def start_kernel(self, kernel_name=None, **kwargs):
    """ Record call and defer to superclass """

class SyncMKMSubclass(RecordCallMixin, MultiKernelManager):
    @mkm_recorder
    def start_kernel(self, kernel_name=None, **kwargs):
    """ Record call and defer to superclass """
```

Thanks.

Reply:
I think something like that will do:
```python
class MKMSubclass(RecordCallMixin):

    @mkm_recorder
    def get_kernel(self, kernel_id):
        """ Record call and defer to superclass """

    # add all the other common methods here

class SyncMKMSubclass(MKMSubclass, MultiKernelManager):

    def _kernel_manager_class_default(self):
        return 'jupyter_client.tests.utils.SyncKernelManagerSubclass'

class AsyncMKMSubclass(MKMSubclass, AsyncMultiKernelManager):

    def _kernel_manager_class_default(self):
        return 'jupyter_client.tests.utils.AsyncKernelManagerSubclass'
```

Regarding the decorator, this looks fine.